### DISCUSSION TITLE
MyPy WIP / research

### DISCUSSION THREAD

**Discussion by jwhonce** [mm21/trilium-alchemy#10] [Ideas]
The following pattern has me stumped regarding typing hints and mypy. 

```python
    note_id: str | None = EntityIdDescriptor()
```

Using the `dataclass.field()` example, I could create a factory function to create `EntityIdDescriptor()` objects. But it doesn't scale well over all the descriptors. It also works by removing the type hint.

```python
def entity_id_descriptor() -> Any:
    return EntityIdDescriptor()
```

I reviewed https://adamj.eu/tech/2021/10/18/python-type-hints-how-to-type-a-descriptor/ as an additional example, but it gets tangled up with the field types defined from the Mixin. Again the way I found to resolve this issue was removing typing hints.

**Comment by mm21** [comment]
Thanks for looking into this! I figured the descriptors were going to be tricky to type hint correctly.

Would mind sharing your branch? I'd be interested in seeing the full MyPy output and maybe trying some experiments. 

Handling `Mixin` might be especially tricky since those attributes shared with `Note` (`note_id`, `title`, `note_type`, `mime`) only exist to provide documentation of how to use it, and they're intended to get "clobbered" by the descriptors in `Note`. In fact if an attribute like `note_id` is defined in a `Mixin`, it will get renamed as `note_id_` when the class is being created (via the metaclass `Meta`). This allows `Note` or `Mixin` subclasses to declaratively define these attributes while keeping the descriptors intact. Later `Note` checks if e.g. `note_id_` is defined, and uses that value if so.

**Comment by jwhonce** [comment]
See https://github.com/mm21/trilium-alchemy/pull/11 for my research/attempt.
